pilot/pkg/networking/core/v1alpha3/envoyfilter/listener_patch.go

```diff
@@ -448,7 +448,7 @@ func doHTTPFilterListOperation(patchContext networking.EnvoyFilter_PatchContext,
 				copy(hcm.HttpFilters[insertPosition+1:], hcm.HttpFilters[insertPosition:])
 				hcm.HttpFilters[insertPosition] = clonedVal
 			}
-		} else if cp.Operation == networking.EnvoyFilter_Patch_INSERT_BEFORE {
+		} else if cp.Operation == networking.EnvoyFilter_Patch_INSERT_BEFORE || cp.Operation == networking.EnvoyFilter_Patch_INSERT_FIRST {
```

Comment:
The comment says that insert-before without a filter match is same as insert first.
That means on line 453 add ‘| insert-before’ condition.
This will exit early.

Reply:
If there is no match, then Patch_INSERT_BEFORE and Patch_INSERT_FIRST both insert the filter at the start, but if there is a match, then you need to find where to insert it.

Reply:
The logic of this block is consistent with the [Network Filter](https://github.com/istio/istio/blob/master/pilot/pkg/networking/core/v1alpha3/envoyfilter/listener_patch.go#L278) processing.

Reply:
The comments is wrong, I will modify it again.

"// insert before/first without a filter match is same as insert in the beginning"

Reply:
I see that network filter is doing the same.

As it standard right now, if a matching position is **not found**, then line 467 block will skip inserting the filter.
That is not the right behaviour for `INSERT_FIRST` of an http filter.

Look at line 423,  `networking.EnvoyFilter_Patch_ADD` clause.
I think we just need a one liner like that to insert in the beginning of `hcm.HttpFilters`.

We could do the same for network filters, but I will need to look at it more closely.

Reply:
@ramaraochavali ^^

Reply:
@ramaraochavali ping you
@mandarjog If there is a match, then INSERT_FIRST is the only way to do it, rather than insert it in any case.

```
INSERT_FIRST | Insert operation on an array of named objects. This operation is typically useful only in the context of filters or routes, where the order of elements matter. Routes should be ordered based on most to least specific matching criteria since the first matching element is selected. For clusters and virtual hosts, order of the element in the array does not matter. Insert first in the list based on the presence of selected filter or not. **This is specifically useful when you want your filter first in the list based on a match condition specified in Match clause**.
```